### ISSUE TITLE
[ENH] Add function to `BaseSeriesAnnotator` to convert from sparse to dense format.

### ISSUE THREAD

**Issue by Alex-JG3** [sktime/sktime#5553]
**Is your feature request related to a problem? Please describe.**
In the annotations module, the `fmt` for `ClaSPSegmentation` can either be `sparse` or `dense` depending on what the user decides when the object is created. This means that depending on `fmt`, the output from `_predict` can either be dense or sparse.

https://github.com/sktime/sktime/blob/9733c36128b8d0b3b5436d4b61597bcdf716cba6/sktime/annotation/clasp.py#L249-L272

**Describe the solution you'd like**

The change points for  `ClaSPSegmentation` are stored as in the `found_cps` attribute in a sparse format. Could we therefore:

* Change the `fmt` for `ClaSPSegmentation` to be `sparse`.
* Remove the `fmt` checking in `_predict` and `_predict_scores`.
* Move the `_get_interval_series` method into `BaseSeriesAnnotator` so other algorithms can use it (like `InformationGainSegmentation`).

This would mean the user would have to manually convert from a `sparse` format to a `dense` format using `_get_interval_series`.

**Describe alternatives you've considered**

**Additional context**

**Comment by Alex-JG3** [comment]
The `ClaspSegmentation` algorithm allows users to change the `fmt` attribute upon instantiation of the model. The `fmt` can either be `sparse` or `dense`. This means that if a user wants to change the output format of their model from `sparse` to `dense` they have to change the `fmt` attribute of their model. For example,

```python
>>> import numpy as np
>>> from sktime.annotation import clasp
>>> n = 50
>>> data = np.concatenate([np.random.normal(-20, 0.1, n), np.random.normal(20, 0.1, n)])
>>> model = clasp.ClaSPSegmentation(fmt="dense")
>>> model.fit(data)
>>> model.predict(data)
IntervalIndex([(0, 55], (55, 100]], dtype='interval[int64, right]')
>>> model.fmt = "sparse"
>>> model.predict(data)
0    55
dtype: int64
```

@fkiraly, are base class attributes changed upon and after model instantiation like this elsewhere in sktime? If not, perhaps we could allow a list of options for the `fmt` attribute in the `BaseSeriesAnnotator` class and then the output format could be selected in the `predict` method? So we add a `fmt` argument to the `predict` method if there are multiple options.

Also, the clasp algorithm seems to use a different definition of  `dense`. From the documentation, I interpret `dense` to mean the output from `predict` should have the same length as `X`.

Here is the relevant documentation.

https://github.com/sktime/sktime/blob/404050dc928bf3899538912778463750f2466ef9/sktime/annotation/base/_base.py#L35-L38

**Comment by fkiraly** [comment]
@Alex-JG3, a timely question. The `fmt` arg, in the design, was always one with which caused me major headache.

There are some conflicting lines of thought I can see:

* if it is an arg - and only then - it can be tuned natively via grid search and the like. This would make sense only if the data container type remains the same, and the "abstract content" changes.
* as you say, you don't really need it in `__init__`, it could be a `predict` arg. However, this would mean all annotation algorithms need to have it.

There are also multiple options we still could take:

* `__init__` arg
* config, via `set_config`, like `sklearn` allows users to set the `predict` output type
* `predict` arg, in this case all, or all estimators of a certain type need to have it do comply with the strategy pattern
* different versions of predict, e.g., `predict_dense` etc, similar to `predict_interval` and `predict_quantiles`.

In the original design, I went with this option, since there are similar instances of special parameters, like `random_state` (or `index` / `columns` now in the distribution design).

Maybe it's a good time to think about pros/cons again. The converter that this issue is looking for will be useful in either case.

**Comment by fkiraly** [comment]
PS: have you seen the pdf with project ideas, in the discord / meetup channel?

**Comment by Alex-JG3** [comment]
> how many of the current annotation algorithms naturally support different fmt options - all, or only some?

Here is a quick summary of the `fmt` and `labels` for each of the annotation algorithms.

`GreedyGaussianSegmentation`
* `fmt` set to `dense` and `labels` is set to `int_label`.
* `fmt` and `labels` are not arguments of `__init__`.
* There is no method for changing between dense and sparse formats.

`InformationGainSegmentation`
* This algorithm uses the default value for `fmt` and `labels`. So, `fmt` is set to `dense` and `labels` is set to `indicator`.
* `fmt` and `labels ` are not an arguments of `__init__`.
* Does not have a method for changing between sparse and dense formats but it inherits from the `SegmentationMixin` class which provides functions for converting change points (e.g. `[0, 0, 1, 0, 0, 1, 0]`) to clusters (e.g. `[0, 0, 1, 1, 1, 2, 2]`). Using the language of the `BaseSeriesAnnotator`, this is like changing the `labels` argument from `indicator` to `int_labels`.

`ClaSPSegmentation`
* `fmt` is set in `__init__` and can be either `sparse` or `dense`. This changes the output format.
* This algorithm also uses a slightly different definition of `dense` as described in my previous comment https://github.com/sktime/sktime/issues/5553#issuecomment-2002628602

`HMM`
* `fmt` is set to `dense`, and `labels` is set to `int_label`.
* `fmt` and `labels ` are not arguments of `__init__`.
* There is no method for changing format.

`EAgglo`
* This inherits from `BaseTransformer` instead of `BaseSeriesAnnotator`.
* This is a change point detection algorithm. If it were to inherit from `BaseSeriesAnnotator`, `labels` would be `int_label` and `format` would be `dense`.
* There is no way to change between the `dense` and `sparse` formats.

`STRAY`
* This inherits from `BaseTransformer` instead of `BaseSeriesAnnotator`.
* This is an outlier detection algorithm.
* If `STRAY` were to inherit from `BaseSeriesAnnotator`, `labels` would be `indicator` and `fmt` would be `dense`.

# My thoughts on the potential ways forward

> `__init__` arg.

I am not a fan of this. I see the parameters passed to `BaseSeriesAnnotator` as private or immutable in some way and should not be changed by users. Please correct me if this is commonly done elsewhere in sktime. 

> config, via `set_config`, like sklearn allows users to set the predict output type, or sktime transformer.

I like this. It seems fairly simple and sounds like a good way to differentiate between configurable parameters and private/immutable parameters in the `BaseSeriesAnnotator` class

> predict arg, in this case all, or all estimators of a certain type need to have it do comply with the strategy pattern

I don't like this. It sounds like it requires a lot of changes where there are potentially simpler solutions.

> different versions of predict, e.g., `predict_dense` etc, similar to `predict_interval` and `predict_quantiles`.

I like this. I think it is very explicit and quite simple.

I am torn between options 2 and 4.

> PS: have you seen the pdf with project ideas, in the discord / meetup channel?

Yes I have just seen it. Thanks for pointing this out.

**Comment by Alex-JG3** [comment]
This has been addressed in  #6265 so closing this.
